protocol/contracts/beanstalk/silo/ConvertFacet.sol

```diff
@@ -77,32 +128,326 @@ contract ConvertFacet is ReentrancyGuard {
+            // Calculate how far past peg we went - so actually this is just abs of new deltaB
+            crossoverAmount = uint256(abs(int256(afterDeltaB)));
+
+            // Check if the crossoverAmount is greater than or equal to bdvConverted
+            // TODO: see if we can find cases where bdcConverted doesn't match the deltaB diff? should always in theory afaict
+            if (crossoverAmount > bdvConverted) {
+                // If the entire bdvConverted amount crossed over, something is fishy, bdv amounts wrong?
+                revert("Convert: converted farther than bdv");
+            } else {
+                return crossoverAmount;
+            }
+        } else if (beforeDeltaB <= 0 && afterDeltaB < beforeDeltaB) { 
+            return amountAgainstPeg;
+        } else if (beforeDeltaB >= 0 && afterDeltaB > beforeDeltaB) { 
+            return amountAgainstPeg;
+        }
+
+        // at this point we are converting in direction of peg, but we may have gone past it
+
+        // Setup convert power for this block if it has not already been setup
+        if (s.convertPowerThisBlock[block.number].hasConvertHappenedThisBlock == false) {
+            // use capped deltaB for flashloan resistance
+            s.convertPowerThisBlock[block.number].convertPower = cappedDeltaB;
+            s.convertPowerThisBlock[block.number].hasConvertHappenedThisBlock = true;
+        }
+
+        // calculate how much deltaB convert is happening with this convert
+        uint256 convertAmountInDirectionOfPeg = abs(beforeDeltaB - afterDeltaB);
+
+        if (convertAmountInDirectionOfPeg <= s.convertPowerThisBlock[block.number].convertPower) {
+            // all good, you're using less than the available convert power
+
+            // subtract from convert power available for this block
+            s.convertPowerThisBlock[block.number].convertPower -= convertAmountInDirectionOfPeg;
+
+            return crossoverAmount;
+        } else {
+            // you're using more than the available convert power
+
+            // penalty will be how far past peg you went, but any remaining convert power is used to reduce the penalty
+            uint256 penalty = convertAmountInDirectionOfPeg - s.convertPowerThisBlock[block.number].convertPower;
+
+            // all convert power for this block is used up
+            s.convertPowerThisBlock[block.number].convertPower = 0;
+
+            return penalty+crossoverAmount; // should this be capped at bdvConverted?
+        }
+    }
+
+    /**
+     * @notice Returns currently availalbe convert power for this block
+     * @return convertPower The amount of convert power available for this block
+     */
+    function getConvertPower() public view returns (uint256) {
+        AppStorage storage s = LibAppStorage.diamondStorage();
+        if (s.convertPowerThisBlock[block.number].hasConvertHappenedThisBlock == false) {
```

Comment:
Since flipping things around this now returns capacity used: https://github.com/BeanstalkFarms/Beanstalk/commit/def87de5c551168fd7901a66dad6e99206cdd8da#diff-80dee2bf35b86cb246d17e5c125566d999c1c76f695851906d62df8db8930950R351